src/dune_rules/gen_rules.ml

```diff
@@ -405,29 +413,34 @@ let rec under_melange_emit_target ~dir =
       | None -> under_melange_emit_target ~dir:parent
       | Some stanza -> Memo.return @@ Some { stanza_dir = parent; stanza }))
 
-let melange_emit_rules sctx { stanza_dir; stanza } =
+let rules_for ~dir components rules =
```

Comment:
I find the way this is factored is making less clear. In some functions, we pass `components`, elsewhere, we use this `components` to compute the `automatic_subdirs_map`. How about we just compute `automatic_subdirs_map` right away and pass that around.

Reply:
check [33a7336](https://github.com/ocaml/dune/pull/7207/commits/33a7336ae9877706f66e693382a96b98a2384933) where I implemented your suggestion.